### PR TITLE
Add microbenchmarks for multiple implementations of BlockStore

### PR DESCRIPTION
**Pull request by YangchenYe323**
### What changes are proposed in this pull request?
Added a benchmark suite to compare the performance between `MonoBlockStore` and `PagedBlockStore`

### Why are the changes needed?
Provide insights on the micro performance characteristic for the `PagedBlockStore` that is under development.

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

**Comment by YangchenYe323** [comment]
@dbw9580 Would you like to take a look at this? Thanks!

**Comment by YangchenYe323** [comment]
Screen Shot 2022-08-10 at 9 08 06 PM

An experimental run on my laptop.

File: microbench/src/main/java/alluxio/worker/BlockStoreReadBench.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+   * A mock consumer of the data read from the store.
+   */
+  private static final byte[] SINK = new byte[MAX_SIZE];
+
+  @State(Scope.Benchmark)
+  public static class BenchParams {
+    private final Random mRandom = new Random();
+
+    @Param({"16", "64"})
+    public long mBlockSizeMB;
```
Comment (dbw9580) [inline microbench/src/main/java/alluxio/worker/BlockStoreReadBench.java:64]: For paged block store, we also want to see the impact of page sizes on read performance.

File: microbench/src/main/java/alluxio/worker/BlockStoreBase.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private static MonoBlockStore createMonoBlockStore(UfsManager ufsManager) {
+    // set up configurations
+    Configuration.set(PropertyKey.WORKER_TIERED_STORE_LEVELS, 1);
+    String tieredStoreDir = AlluxioTestDirectory
+        .createTemporaryDirectory("worker_local_store").getAbsolutePath();
+    Configuration.set(PropertyKey.WORKER_TIERED_STORE_LEVEL0_DIRS_PATH, tieredStoreDir);
+
+    // mock a BlockMasterClientPool as we won't use it
```
Comment (dbw9580) [inline microbench/src/main/java/alluxio/worker/BlockStoreBase.java:72]: it's actually used in `commitBlock`, but that's during preparation and does not affect actuall measurement iterations, so is fine.

**Comment by dbw9580** [comment]
> Screen Shot 2022-08-11 at 12 35 35 AM
> An experimental run on my laptop.

I wonder why `monoBlockStoreReadUfs` is significantly slower than any other ones?

**Comment by dbw9580** [comment]
https://github.com/Alluxio/alluxio/blob/8f74e6584b6d2d71be3a3737057808b6f77c44ae/core/common/src/main/java/alluxio/worker/block/io/LocalFileBlockReader.java#L100

javadoc on `java.nio.channels.FileChannel#map`says

> For most operating systems, mapping a file into memory is more expensive than reading or writing a few tens of kilobytes of data via the usual read and write methods. From the standpoint of performance it is generally only worth mapping relatively large files into memory.

looks like memory-mapping a file has a significant overhead... even for a file as large as 64MB.

**Comment by jja725** [comment]
It's surprising that monoBlockStoreReadLocal performance is so bad...

**Comment by YangchenYe323** [comment]
https://lkml.indiana.edu/hypermail/linux/kernel/0802.0/1496.html There's been discussions about performance issues of `mmap`. In general it seems that `mmap` brings significant overhead with like additional page fault, and could easily end up slower than normal `read` calls.

**Comment by YangchenYe323** [comment]
Screen Shot 2022-08-11 at 2 37 04 PM

Experimental run with page size as parameters.

**Comment by YangchenYe323** [comment]
Screen Shot 2022-08-12 at 12 06 56 AM

Re-benchmarked MonoBlockStore using ramdisk as local storage. I'll make a compilation of all the results in a spreadsheet.

**Comment by dbw9580** [comment]
I take a closer look at the mmap performance on Linux.

图片

Looks like there is quite a significant difference in perf between Linux and Mac.

My microbench is here https://gist.github.com/dbw9580/2bdb08ea6bf8e44e94a7f28c2d06155d

**Comment by YangchenYe323** [comment]
> I take a closer look at the mmap performance on Linux.
> 
> 图片
> 
> Looks like there is quite a significant difference in perf between Linux and Mac.
> 
> My microbench is here https://gist.github.com/dbw9580/2bdb08ea6bf8e44e94a7f28c2d06155d

Yeah my running suggests memory mapping is a lot slower on MacOS. Also, the current `UnderFileSytemBlockReader` uses a `BufferedInputStream` under the hood rather than a `RandomAccessFile`. That might be why `readUfs` could still be faster than `readLocal`.

**Comment by YangchenYe323** [comment]
Added an experimental run of Random Read results.

**Comment by beinan** [comment]
Great job!  just post two flamecharts for paged read and transfer.
CPU-flame-graph-pooled-transfer
CPU-flame-graph-unpooled-read

**Comment by YangchenYe323** [comment]
Benchmark results on Linux. Seems that MacOS does have particular issues with `mmap`
Screen Shot 2022-08-17 at 1 01 10 AM
Screen Shot 2022-08-17 at 1 01 28 AM

**Comment by YangchenYe323** [comment]
@dbw9580 Do you think this pr is good to merge now?

**Review by dbw9580 — APPROVED** [review]
LGTM

**Review by beinan — APPROVED** [review]
Lgtm, indeed I have been using this for a while, great work!

**Comment by beinan** [comment]
alluxio-bot, merge this please

**Pull request merged**